### ISSUE TITLE
Subtract with Overflow when using credit-only accounts

### ISSUE THREAD

**Issue by sagar-solana** [solana-labs/solana#4958]
#### Problem

Observed a CI failure in `bench::tests::test_exchange_local_cluster` pointing to a subtract with overflow related to credit-only accounts. 

#### Proposed Solution

Needs triage

**Comment by sagar-solana** [comment]
CI failed while running `test-stable.sh`, specifically `bench::tests::test_exchange_local_cluster`.
This error seems flaky, so maybe a race somewhere? 


```
thread 'solana-banking-stage-tx' panicked at 'attempt to subtract with overflow', runtime/src/accounts.rs:411:25
  | note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
  | stack backtrace:
  | 0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
  | at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
  | 1: std::sys_common::backtrace::_print
  | at src/libstd/sys_common/backtrace.rs:71
  | 2: std::panicking::default_hook::{{closure}}
  | at src/libstd/sys_common/backtrace.rs:59
  | at src/libstd/panicking.rs:197
  | 3: std::panicking::default_hook
  | at src/libstd/panicking.rs:211
  | 4: std::panicking::rust_panic_with_hook
  | at src/libstd/panicking.rs:474
  | 5: std::panicking::continue_panic_fmt
  | at src/libstd/panicking.rs:381
  | 6: rust_begin_unwind
  | at src/libstd/panicking.rs:308
  | 7: core::panicking::panic_fmt
  | at src/libcore/panicking.rs:85
  | 8: core::panicking::panic
  | at src/libcore/panicking.rs:49
  | 9: solana_runtime::accounts::Accounts::unlock_account
  | at runtime/src/accounts.rs:411
  | 10: solana_runtime::accounts::Accounts::unlock_accounts::{{closure}}
  | at runtime/src/accounts.rs:488
  | 11: core::iter::traits::iterator::Iterator::for_each::{{closure}}
  | at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libcore/iter/traits/iterator.rs:604
  | 12: core::iter::traits::iterator::Iterator::fold::{{closure}}
  | at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libcore/iter/traits/iterator.rs:1684
  | 13: core::iter::traits::iterator::Iterator::try_fold
  | at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libcore/iter/traits/iterator.rs:1572
  | 14: core::iter::traits::iterator::Iterator::fold
  | at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libcore/iter/traits/iterator.rs:1684
  | 15: core::iter::traits::iterator::Iterator::for_each
  | at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libcore/iter/traits/iterator.rs:604
  | 16: solana_runtime::accounts::Accounts::unlock_accounts
  | at runtime/src/accounts.rs:487
  | 17: solana_runtime::bank::Bank::unlock_accounts
  | at runtime/src/bank.rs:763
  | 18: <solana_runtime::locked_accounts_results::LockedAccountsResults as core::ops::drop::Drop>::drop
  | at runtime/src/locked_accounts_results.rs:39
  | 19: core::ptr::real_drop_in_place
  | at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libcore/ptr.rs:193
  | 20: core::mem::drop
  | at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libcore/mem.rs:777
  | 21: solana::banking_stage::BankingStage::process_and_record_transactions
  | at core/src/banking_stage.rs:541
  | 22: solana::banking_stage::BankingStage::process_transactions
  | at core/src/banking_stage.rs:574
  | 23: solana::banking_stage::BankingStage::process_received_packets
  | at core/src/banking_stage.rs:717
  | 24: solana::banking_stage::BankingStage::process_packets
  | at core/src/banking_stage.rs:825
  | 25: solana::banking_stage::BankingStage::process_loop
  | at core/src/banking_stage.rs:370
  | 26: solana::banking_stage::BankingStage::new_num_threads::{{closure}}::{{closure}}

```

**Comment by sagar-solana** [comment]
Ref: https://github.com/solana-labs/solana/blob/ef7022d638b077b715933148cbc52837f77b54cc/runtime/src/accounts.rs#L410-L412

**Comment by rob-solana** [comment]
another occurrence:
https://buildkite.com/solana-labs/solana/builds/8772#d2851747-42aa-4796-8a4d-2bea3ea3604c

**Comment by CriesofCarrots** [comment]
Appears to be resolved. Please re-open if new incidences occur.

Closed #5028
